rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+  }:
+  {
+    result = x + y;
+  };
+```
+
+Notable special cases are:
+
+- Single line values that would not benefit from style #2 keep using #1, even if this makes it go above the line limit. This mostly applies to simple strings and paths.
+- Attribute set values are *always* expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
+  ```nix
+  {
+    foo.bar.baz = "qux";
+    foo' = {
+      bar.baz = "qux";
+    };
+  }
+  ```
+- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
+- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
+  ```nix
+  buildInputs = with pkgs; [
+    some
+    dependencies
+  ];
+  ```
+
+**Alternatives**
+
+One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binder is very long:
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+#### inherit
+
+**Description**
+
+The items are either all on the same line, or all on a new line each (with indentation).
+
+**Examples**
+
+```nix
+inherit foo bar baz;
+inherit
+  foo'
+  bar'
+  baz'
+;
+```
+
+#### inherit from
+
+**Description**
+
+- If the inherit target is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
+- Otherwise, it starts on a new line with indentation, like the others.
+  - In that case, the remaining items are force-expanded too, even if they would have fit onto one line in the first place.
+
+**Examples**
+
+```nix
+inherit (pkgs) ap1 ap2 ap3;
+inherit (pkgs)
+  app1
+  app2
+  # …
+  app42
+;
+inherit
+  (pkgs.callPackage ./foo.nix {
+    arg = "val";
+  })
+  attr1
+  attr2
+;
+```
+
+#### Semicolon placement
+
+**Description**
+
+The semicolon is always placed on the same line as the expression it concludes.
+
+**Examples**
+
+```nix
+{
+  attr1 = bar;
+  attr2 = function call {
+    # stuff
+  };
+  attr3 =
+    function call
+      many
+      arguments;
+  attr4 =
+    let
+      foo = "bar";
+    in
+    some statement;
+  attr5 =
+    if foo then
+      "bar"
+    else
+      "baz";
+  attr6 =
+    let
+      foo = false;
+    in
+    if foo then "bar" else "baz";
+  attr7 = function (
+    if foo then
+      "bar"
+    else
+      "baz"
+  );
+  attr8 =
+    cond1
+    || cond2
+    ||
+      some function call
+      && cond3;
+}
+```
+
+**Rationale and alternatives**
+
+
+There are four considered semicolon styles:
+1. On a new line without indentation.
+  - This clearly marks a separation between attributes, however it is wasteful of space.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments;
+  attr3 =
+    let
+      foo = "bar";
+    in
+    some statemens;
```

Comment:
That would match the description:

```suggestion
  attr3 =
    function call
      many
      arguments
  ;
  attr3 =
    let
      foo = "bar";
    in
    some statemens
  ;
```

Reply:
Oh yeah we accidentally formatted these lines even though they're in the alternatives, fixed in https://github.com/NixOS/rfcs/pull/166/commits/63a2fa906bbf5929e9b4d82b7047619e39ea718f.